eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1327 @@
+       ++parent_featured_sets_index)
+    {
+      EknrBoxedSetInfo *set = g_ptr_array_index (parent_featured_sets,
+                                                 parent_featured_sets_index);
+
+      g_autoptr(JsonBuilder) child_builder = json_builder_new ();
+
+      json_builder_begin_object (child_builder);
+
+      /* Build up JSON representation of child */
+      json_builder_set_member_name (child_builder, "child_tags");
+      json_builder_add_value (child_builder,
+                              strv_to_json_array_node ((const char * const *) set->child_tags));
+
+      json_builder_set_member_name (child_builder, "ekn_id");
+      json_builder_add_string_value (child_builder, set->ekn_id);
+
+      json_builder_set_member_name (child_builder, "title");
+      json_builder_add_string_value (child_builder, set->title);
+
+      json_builder_set_member_name (child_builder, "tags");
+      json_builder_add_value (child_builder,
+                              strv_to_json_array_node ((const char * const *) set->tags));
+
+      json_builder_end_object (child_builder);
+
+      /* Add to the array */
+      json_builder_add_value (array_builder,
+                              json_builder_get_root (child_builder));
+    }
+
+  json_builder_end_array (array_builder);
+
+  /* Set parent object value */
+  json_builder_add_value (object_builder, json_builder_get_root (array_builder));
+  json_builder_end_object (object_builder);
+
+  /* Stringify */
+  node = json_builder_get_root (object_builder);
+  generator = json_generator_new ();
+
+  json_generator_set_root (generator, node);
+  json_string = json_generator_to_data (generator, NULL);
+
+  return g_variant_new_string (json_string);
+}
+
+static GVariant *
+get_wrapper_metadata (const char         *ekn_id,
+                      const char         *title,
+                      const char         *published,
+                      const char * const *authors,
+                      const char         *license,
+                      const char         *source,
+                      const char         *source_name,
+                      const char         *original_uri,
+                      const char * const *tags,
+                      GHashTable         *set_map)
+{
+  g_autoptr(JsonBuilder) metadata_builder = json_builder_new ();
+  g_autoptr(JsonBuilder) sets_array_builder = json_builder_new ();
+  g_autoptr(GPtrArray) sets = g_ptr_array_new_with_free_func ((GDestroyNotify) eknr_boxed_set_info_unref);
+  guint sets_index = 0;
+  g_autoptr(JsonNode) node = NULL;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autofree char *json_string = NULL;
+
+  for (; *tags != NULL; ++tags)
+    {
+      const char *tag = *tags;
+      EknrBoxedSetInfo *set = NULL;
+
+      if (strncmp (tag, "Ekn", 3) == 0)
+        continue;
+
+      set = g_hash_table_lookup (set_map, tag);
+
+      if (set == NULL)
+        continue;
+
+      g_ptr_array_add (sets, eknr_boxed_set_info_ref (set));
```

Comment:
There's no need to ref/unref the sets when adding them to the array; `set_map` will stay alive and won't change during the execution of this function.

Reply:
True that - thanks!